sysusers: split up systemd.conf

calestyo opened the pull request:
This makes it easier have the respective users/groups only created when their
respective packages are installed.

Fixes #20044.

yuwata commented on sysusers.d/systemd-coredump.conf.in:
```diff
@@ -0,0 +1,10 @@
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+{% if ENABLE_COREDUMP %}
```
This condition seems not necessary any more.

calestyo replied:
I've had already asked for it in https://github.com/systemd/systemd/issues/20044#issuecomment-869876506 ...

I guess one can drop it then from `systemd-remote.conf.in`, too? That's why I kept it originally, cause that had it, too.

yuwata commented on sysusers.d/systemd-network.conf.in:
```diff
@@ -0,0 +1,10 @@
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+{% if ENABLE_NETWORKD %}
```
Also here.

mbiebl replied:
Right, there doesn't need to be a .in pre-processing anymore.

mbiebl replied:
argh, for networkd, you still need .in pre-processing for the UID, but no condition, as @yuwata said.

So disregard what I said. I was looking at systemd-oom.conf.in when commenting.

calestyo replied:
> argh, for networkd, you still need .in pre-processing for the UID, but no condition, as @yuwata said.

So the `{% if ENABLE_NETWORKD %}` around is needed for the inner `{{SYSTEMD_NETWORK_UID}}` to work?

keszybz replied:
No.

yuwata reviewed (changes requested):
I guess conditions in each .conf are not necessary.

mbiebl commented:
The ubuntu CI failures are expected. If/Once this PR is merged, I'd update the packaging files in the upstream-ci branch to unbreak this again.

poettering commented:
I am not sure how much I like this. Splitting this up means people's overrides and extends don't work anymore. The file names this stuff is listed in kinda is an API

poettering commented:
What's the benefit of this? people being allergic to a bunch of extra UIDs being allocate? is that really something the trouble of breaking API for? Admittedly it's a weak for API, but it's effectively API nonetheless, and we *did* have reports about this when we did something similar in earlier cases.

mbiebl commented:
> What's the benefit of this?

I'm a bit torn on this. My first reaction was similar to yours. Contemplating a bit more on it, I see two possible benefits:
- For distros that split up systemd, it's as @keszybz said, more elegant / conceptually cleaner that the invididual package declares what user it needs/uses. 
- It makes it easier to override/extend individual entries. If you have a single/monolithic systemd.conf and you only want to change the settings for say systemd-network, you'll have to override all of systemd.conf. This means, whenever systemd.conf changes, you have to review those changes and adapt accordingly if necessary.

You raise a valid concern though, that it might break existing systems, which have an override for the monolithic systemd.conf.

At least for Debian/Ubuntu, we could cover that, by adding an appropriate preinst check. If such an `/etc/sysusers.d/systemd.conf` exists, we could output a warning.

mbiebl commented:
> Admittedly it's a weak for API, but it's effectively API nonetheless, and we _did_ have reports about this when we did something similar in earlier cases.

Maybe I missed that, but what was this about exactly? Maybe this is something we can address in our Debian packages with a corresponding upgrade check.

keszybz commented:
The "api break" is negligible. The easiest and most obvious way to "override" the sysusers settings is to precreate the user and/or group. I think it's very unlikely that this would create any problems. We did such splits with tmpfiles before, and there it's more likely for people to create overrides, and I don't recall any complains. We'd just add a NEWS entry and things will be fine.

calestyo commented:
I've updated the first commit to remove the meson stuff, and added a 2nd one which also removes it for the already separate `systemd-remote.conf.in`.

Felt that didn't belong into one commit, as for `systemd-remote.conf.in` it's not really about splitting the file up, but just removing what was (AFAIU) already not needed without the split.

yuwata commented:
So, now, some of them are not necessary to be processed with jinja2, and `.in` suffix can be dropped.

calestyo commented:
> So, now, some of them are not necessary to be processed with jinja2, and `.in` suffix can be dropped.

updated accordingly

yuwata commented:
Thank you for the quick update. But meson file also needs to be updated to handle files without the suffix correctly.

calestyo commented:
I've never really used meson... how do I conditionally include the non-`.in` files? Or at least I think they should still be only included if the respective flag is set?

keszybz commented:
>  how do I conditionally include the non-`.in` files? Or at least I think they should still be only included if the respective flag is set?

The `tmpfiles.d/meson.build` file has a good example.

calestyo commented:
Sorry, for needing help again, but when I closely follow what's done for tmpfiles.d, e.g. something like:
```
files = [['README'], ''],
         ['systemd-coredump.conf', 'ENABLE_COREDUMP'],
         ['systemd-oom.conf',      'ENABLE_OOMD'],
         ['systemd-remote.conf',   'ENABLE_REMOTE']]

foreach pair : files
        if not enable_sysusers
                # do nothing
        elif pair[1] == '' or conf.get(pair[1]) == 1
                install_data(pair[0], install_dir : sysusersdir)
        else
                message('Not installing sysusers.d/@0@ because @1@ is @2@'
                        .format(pair[0], pair[1], conf.get(pair[1], 0)))
        endif
endforeach
```

Than I obviously have the problem that I cannot really fit the `and conf.get('HAVE_MICROHTTPD') == 1]]` into it.
Is there either some nice way in meson to make the 2nd element of the pair another array that's iterated over, or shall I simply included the boolean outcome of the condition in `files` and take that?

keszybz commented:
> shall I simply included the boolean outcome of the condition in `files` and take that?

Since there's just one file, you can do something like this:
```diff
  if enable_sysusers
-         install_data('README', install_dir : sysusersdir)
+         install_data('README',
+                      'systemd-remote.conf',
+                      install_dir : sysusersdir)
  endif
```

calestyo commented:
Okay check please check these commits if they fit.

calestyo commented:
Anything left to do on this?

keszybz commented:
Sorry for the delay. Looks great, thanks.
